### ISSUE TITLE
Outlet Reservation Support

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#1958]
_Imported from trac ticket 1958.
Created by **kbower**
Opened in trac: Sep 29, 2012, 8:10:43 AM
Last modified in trac: Feb 14, 2013, 9:18:57 AM_
***
Support for Reserving outlet.

See also [#2090](2090) and [#2074](2074) 

***
* kbower attached [version12 reserve outlet.txt](https://trac.retailarchitects.com/trac/attachment/ticket/1958/version12 reserve outlet.txt) on Feb 14, 2013, 9:19:56 AM
* kbower attached [version10 reserve outlet.txt](https://trac.retailarchitects.com/trac/attachment/ticket/1958/version10 reserve outlet.txt) on Feb 14, 2013, 9:20:09 AM

**Comment by kentbower** [comment]
_**sjohnson** trac comment on Jan 11, 2013, 9:54:11 AM:_

Morris Uses Outlet.  Need to be able to support taking outlet orders in Loft.

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 11, 2013, 3:23:14 PM:_

go live blocker

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 17, 2013, 9:41:08 AM:_

Dividing this ticket for speed of implementation.  See [#2074](2074) for NAS `LockedInventory` support.

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 17, 2013, 11:11:43 AM:_

Reserved inventory _is_ deleted from OUT.

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 17, 2013, 4:16:51 PM:_

Darlene,

1. Is OUT_COMM_CD a commission code?  How does that work?  Is that
   simply moved to the SO_LN when reserved?
2. Is SPIFF moved to the SO_LN.SPIFF when reserved?
3. I assume ENTRY_DT is moved to SO_LN.OUT_ENTRY_DT?
4. Are CRPT_ID_NUM and CRPT_SER_NUM stored on SO_LN?  Where?
5. What about CMNT?  Is that stored on SO_LN?  Where?

Thanks!
Kent

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 18, 2013, 8:39:36 AM:_

Darlene,
    The NAS_CD and OUT_CD tables have a SUPPRESS column:

SQL> select * from nas_cd;

NAS DES                                      S
--- ---------------------------------------- -
FRT FRIEGHT CLAIM                            N
GWC GUARDIAN CLAIM WAITING CHECK             N
OUT OUTLET                                   N
POR PARTS ON ORDER                           N
NIL NOT IN LOCATION                          Y
RAV REPAIR AT VENDOR                         Y
RFS READY FOR SHOP                           N
NTO NEED TO ORDER PARTS                      N
BTI BUYER TO INSPECT                         N
RTV RETURN TO VENDOR                         N
ADD APPROVED TO DONATE/DESTROY               N
RFO READY FOR OUTLET                         N
COW CUSTOMER OWNED NEEDS REPAIRED            N
STA STAINSAFE RETURNS                        N
TBD TO BE DISPOSITIONED                      N
PIC PICKING EXCEPTIONS                       N
GUA GUARDIAN CLAIMS                          N
PEN WAITING FOR VENDOR CREDIT                N

18 rows selected.

When you make SUPPRESS = 'Y', is that code _disabled_ or just _hidden, but still usable_?

Thank You!
Kent

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 18, 2013, 10:05:06 AM:_

**response in bold**
Darlene,
We've had the understanding/assumption that one of the main benefits to using OUTLET functionality is that you can individually price as-is or scratch-and-dent type clearance pieces.  Apparently this is not the case, which surprised me greatly.  My understanding now is that the main benefit is to be able to track reduced costs. For example, outlet allows you to track a lower cost allowed by a vendor for floor sample pieces.

However, I understood that V12 also allows this, so the natural question now is: 1) why does Morris use outlet?  (And my guess is the reason is "it's been the procedure since before V12," but what I'm really asking is: 2) Are there also other benefits to using OUTLET, and what are they?)**—it is easier to be sure that the reduced retail is going against the correct item if it is put into an outlet status.  It also allows us to reduce the cost on that particular piece if we get additional credits from a vendor on an item. For example, if there are 3 in store 02 not in an outlet status and then key a cost adjustment, there is no way of knowing which item the cost adjustment will go against.  We tag items in the outlet based on reduced cost and physical shape of an item.**

Further, a technical question: 3) Since V12 tracks outlet inventory in both ITM_INV and OUT, and *both* these tables have a CST column, which one is the field GERS honors?  It seems bad database design because you could have conflicting costs in the two tables for the same physical piece.**—I would assume that the cost in OUT is the ruling cost. I tested one of the items below in train but found that the cost is the same in both tables???**

A database query from your live data shows this is the case for 15 of your pieces:
```
  1  select INV_ID, NAS_OUT_ID, i.itm_cd, i.store_cd, i.loc_cd, i.CST as INV_COST, o.CST as OUT_COST
  2  from itm_inv i, out o
  3  where i.INV_STATUS = 'O' and i.NAS_OUT_ID = o.ID_CD
  4  and (i.CST != o.CST or i.cst is null)
  5* order by i.store_cd, i.itm_cd
SQL> /

INV_ID     NAS_OUT_ID ITM_CD    STOR LOC_CD   INV_COST OUT_COST
---------- ---------- --------- ---- ------ ---------- ----------
073012QLXJ 1016279354 381110875 02   OUTLET         22      23.61
102849GZBP 0602162614 422172015 02   OUTLET         50         42
062612JVRJ 0703251945 319897584 11   OUTLET         55         36
121510LGZY 1003186691 382816505 11   OUTLET      37.54      57.75
102631CKCF 1217027771 883001706 11   OUTLET      14.96      14.95
124530IRDS 1217027770 883001706 11   OUTLET      14.96      14.95
020210PCVE 0313225717 890354992 11   OUTLET         51      51.15
111909PCVD 0313225716 890354992 11   OUTLET       51.3      51.15
120512AKTZ 1213298367 200218356 12   OUTLET     179.99     173.32
022211PVFU 1213298369 200218356 12   OUTLET     169.99     173.32
111011QDJL 0125213612 212110768 12   OUTLET      84.99      82.49
070510NOWP 0125213608 212110768 12   OUTLET      79.99      82.49
090512PWGJ 1213298404 212816891 12   OUTLET     104.99     103.99
021312YFEU 1213298405 212816891 12   OUTLET     102.99     103.99
072611JECL 1208102316 467150947 12   OUTLET         67         65

15 rows selected.
```

4) So, if you run your GERS reports, which CST is GERS using, the one in ITM_INV or OUT?
5) If you reserve a piece, which cost does GERS pull to SO_LN.FIFO_CST? (if you don't mind testing this from E1 in TRAIN, I've run the query on TRAIN also below)

Thanks again!
Kent




TRAIN results:
```
INV_ID     NAS_OUT_ID ITM_CD    STOR LOC_CD   INV_COST OUT_COST
---------- ---------- --------- ---- ------ ---------- ----------
012111LSYF 0516241320 339814128 02   OUTLET         80      83.75
112311RFON 0516241323 339814128 02   OUTLET         85      83.75
030812LCVH 0919271149 398122259 02   OUTLET      35.18      35.17
102849GZBP 0602162614 422172015 02   OUTLET         50         42—I cannot see a cost difference in these 2 tables on this item???
062612JVRJ 0703251945 319897584 11   OUTLET         55         36
121510LGZY 1003186691 382816505 11   OUTLET      37.54      57.75
030612ZXDC 0914270047 402032469 11   OUTLET         42         31
124530IRDS 1217027770 883001706 11   OUTLET      14.96      14.95
102631CKCF 1217027771 883001706 11   OUTLET      14.96      14.95
020210PCVE 0313225717 890354992 11   OUTLET         51      51.15
111909PCVD 0313225716 890354992 11   OUTLET       51.3      51.15
070510NOWP 0125213608 212110768 12   OUTLET      79.99      82.49
111011QDJL 0125213612 212110768 12   OUTLET      84.99      82.49
081111IZYR 0326149056 397137146 12   OUTLET         50       48.5
072611JECL 1208102316 467150947 12   OUTLET         67         65

15 rows selected.
```

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 18, 2013, 3:59:14 PM:_

Do you know whether GERS allows OUT inventory to have qty > 1?

Kent

___
edit: Both Wayne at Kane's and Darlene at Morris report that qty > 1 is not allowed for OUT (which makes one scratch his head and wonder why the database column "QTY" exists on OUT).  This simplifies things greatly, but Loft will be programmed to double check QTY is 1.

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 23, 2013, 7:17:02 AM:_

Darlene, thank you for your reply.

You said you tag the items based on reduced cost.  Do you just hand write the price on the tag?  Is there somewhere in the system for entering the assigned outlet price or does the salesperson need to enter it manually on the SO_LN?**—we hand write the tag and the price is entered with over-ride approval**

As for OUT piece 0602162614 in TRAIN below, when you opened it in E1, it apparently "corrected" the cost in OUT to be the same as ITM_INV, because now they both show $50, as you saw:
```
1* select  NAS_OUT_ID, itm_cd, cst from itm_inv where NAS_OUT_ID = '0602162614'

NAS_OUT_ID ITM_CD           CST
---------- --------- ----------
0602162614 422172015         50

SQL> select ID_CD, itm_cd, cst from out where id_cd = '0602162614';

ID_CD      ITM_CD           CST
---------- --------- ----------
0602162614 422172015         50
```
Unless you did that manually?  It's possible E1 takes the greater of the two costs, so can you please also inspect 0516241320 and see whether you see $80 or $83.75?**—I entered a sale for this item and the cost was $87.20????**

Also can you please (without first looking in whatever screen corrects the cost, in other words, by going straight to the order entry screen) write an order for ITM_CD 319897584 and fill OUT ID 0703251945 which is in store 11 OUTLET?
```
SQL> select  NAS_OUT_ID, itm_cd, cst from itm_inv where NAS_OUT_ID = '0703251945';
NAS_OUT_ID ITM_CD           CST
---------- --------- ----------
0703251945 319897584         55

SQL> select ID_CD, itm_cd, cst from out where id_cd = '0703251945';

ID_CD      ITM_CD           CST
---------- --------- ----------
0703251945 319897584         36
```
Please let me know once you've filled this order in TRAIN and send me the DEL DOC NUM, ok?**—so#0123311NNOQ**

Thanks!

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 24, 2013, 2:53:23 PM:_

Your two companies are the only of our clients using GERS OUTLET functionality.  You also are on very different GERS versions, but I'm hoping you can help figure these things out so we can implement OUTLET for Loft.

When outlet inventory is reserved the OUT record is deleted and (most) the data is copied to the SO_LN.

Here are the simple OUT fields:
OUT.ID_CD  => copied to => SO_LN.OUT_ID_CD
OUT.STORE_CD  => copied to => SO_LN.STORE_CD
OUT.LOC_CD => copied to => SO_LN.LOC_CD
OUT.OUT_CD => copied to => SO_LN.OUT_CD
OUT.ITM_CD => copied to => SO_LN.ITM_CD
OUT.CRPT_SER_NUM => copied to => SO_LN.SER_NUM
OUT.OUT_COMM_CD => copied to => SO_LN.COMM_CD

The rest of the OUT fields:
OUT.CRPT_ID_NUM  is presumed copied to SO_LN.ID_NUM?  (Does your company use this?)
OUT.QTY && SO_LN.QTY **MUST BOTH BE 1** according to my understanding, please confirm.

OUT.CST for V12 is apparently ignored and the CST from ITM_INV is copied to SO_LN.FIFO_CST
For V10, I presume OUT.CST is copied straight to SO_LN.FIFO_CST? (Michelle/Sal, please confirm.)

OUT.ENTRY_DT:
Although SO_LN.OUT_ENTRY_DT exists, apparently, at least in V12, OUT.ENTRY_DT is **not** copied, for some reason. Instead, if unreserved, the OUT record gets a new OUT.ENTRY_DT of the date of unfill.  (Michelle/Sal, please confirm for V10)
Does anyone know if there is a reason for this?  Should unreserved OUTLET inventory for some reason appear as "newly" entered OUTLET? It doesn't seem this should be the case, so unless someone knows a reason, Loft will preserve the original outlet entry date by copying it to SO_LN.OUT_ENTRY_DT.

OUT.SPIFF:
At least for V12, is NOT copied to SO_LN.SPIFF.  (Michelle/Sal, please confirm for V10)  Is this correct behavior?  What is the purpose of having OUT.SPIFF?

OUT.CMNT:
Is lost, not copied to SO_LN.  Loft will fix this for inventory reserved and unfilled in Loft, the CMNT will not be lost.

OUT.FIFL_DT:
For V12, is NOT copied to SO_LN.FIFL_DT, instead seems inconsistently copied from ITM_INV.RCV_DT/OUT.FIFL_DT?  Darlene, any ideas there?  Would copying OUT.FIFL_DT to SO_LN.FIFL_DT be acceptable?
Michelle/Sal: for V10, is OUT.FIFL_DT copied straight to SO_LN.FIFL_DT?

OUT.FIFO_DT:
For V12, is not copied to SO_LN.FIFO_DT; instead seems to be taken from ITM_INV.RCV_DT.
Michelle/Sal: for V10, is OUT.FIFO_DT copied straight to SO_LN.FIFO_DT?


Please let me know,
Kent

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 25, 2013, 8:52:55 AM:_

Darlene Replying to [kbower](comment:13):

> OUT.QTY && SO_LN.QTY '''MUST BOTH BE 1''' according to my understanding, please confirm.

--YES

> OUT.ENTRY_DT:
> Although SO_LN.OUT_ENTRY_DT exists, apparently, at least in V12, OUT.ENTRY_DT is '''not''' copied, for some reason. Instead, if unreserved, the OUT record gets a new OUT.ENTRY_DT of the date of unfill.  (Michelle/Sal, please confirm for V10)
> Does anyone know if there is a reason for this?  Should unreserved OUTLET inventory for some reason appear as "newly" entered OUTLET? It doesn't seem this should be the case, so unless someone knows a reason, Loft will preserve the original outlet entry date by copying it to SO_LN.OUT_ENTRY_DT.

—I would think the original date should be preserved.
> OUT.FIFL_DT:
> For V12, is NOT copied to SO_LN.FIFL_DT, instead seems inconsistently copied from ITM_INV.RCV_DT/OUT.FIFL_DT?  Darlene, any ideas there?  Would copying OUT.FIFL_DT to SO_LN.FIFL_DT be acceptable?

—I would think so

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 25, 2013, 1:13:12 PM:_

Michelle @ IvanSmith Replying to [kbower](comment:13):
> Here are the simple OUT fields:
> OUT.ID_CD  => copied to => SO_LN.OUT_ID_CD
> OUT.STORE_CD  => copied to => SO_LN.STORE_CD
> OUT.LOC_CD => copied to => SO_LN.LOC_CD
> OUT.OUT_CD => copied to => SO_LN.OUT_CD
> OUT.ITM_CD => copied to => SO_LN.ITM_CD
> OUT.CRPT_SER_NUM => copied to => SO_LN.SER_NUM
> OUT.OUT_COMM_CD => copied to => SO_LN.COMM_CD

(yes to all above)
> The rest of the OUT fields:
> OUT.CRPT_ID_NUM  is presumed copied to SO_LN.ID_NUM?  (Does your company use this?)

(no)
> OUT.QTY && SO_LN.QTY '''MUST BOTH BE 1''' according to my understanding, please confirm.

(yes) 
> 
> OUT.CST for V12 is apparently ignored and the CST from ITM_INV is copied to SO_LN.FIFO_CST
> For V10, I presume OUT.CST is copied straight to SO_LN.FIFO_CST? (Michelle/Sal, please confirm.)

(yes) 
> 
> OUT.ENTRY_DT:
> Although SO_LN.OUT_ENTRY_DT exists, apparently, at least in V12, OUT.ENTRY_DT is '''not''' copied, for some reason. Instead, if unreserved, the OUT record gets a new OUT.ENTRY_DT of the date of unfill.  (Michelle/Sal, please confirm for V10)

(yes)
> Does anyone know if there is a reason for this?  Should unreserved OUTLET inventory for some reason appear as "newly" entered OUTLET? It doesn't seem this should be the case, so unless someone knows a reason, Loft will preserve the original outlet entry date by copying it to SO_LN.OUT_ENTRY_DT.

(still checking on this)
> 
> OUT.SPIFF:
> At least for V12, is NOT copied to SO_LN.SPIFF.  (Michelle/Sal, please confirm for V10)  Is this correct behavior?  What is the purpose of having OUT.SPIFF?

(OUT.SPIFF is copied to SO_LN.SPIFF in V10. To my knowledge we dont use it but it could be a tool to get outlet out the door quicker) 
> 
> OUT.CMNT:
> Is lost, not copied to SO_LN.  Loft will fix this for inventory reserved and unfilled in Loft, the CMNT will not be lost.

(good) 
> 
> OUT.FIFL_DT:
> For V12, is NOT copied to SO_LN.FIFL_DT, instead seems inconsistently copied from ITM_INV.RCV_DT/OUT.FIFL_DT?  Darlene, any ideas there?  Would copying OUT.FIFL_DT to SO_LN.FIFL_DT be acceptable?
> Michelle/Sal: for V10, is OUT.FIFL_DT copied straight to SO_LN.FIFL_DT?

(yes) 
> 
> OUT.FIFO_DT:
> For V12, is not copied to SO_LN.FIFO_DT; instead seems to be taken from ITM_INV.RCV_DT.
> Michelle/Sal: for V10, is OUT.FIFO_DT copied straight to SO_LN.FIFO_DT?

(yes)

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 25, 2013, 1:14:24 PM:_

Michelle, thanks for the response.

Regarding ENTRY_DT, it seems extremely arbitrary that simply reserving and unreserving an OUTLET piece should change its ENTRY_DT.  That makes no sense to me, so unless I hear otherwise from you, Loft will preserve this date, regardless of being reserved/unreserved.

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 11, 2013, 3:57:33 PM:_

Does one of the 3 of you know where the SO_LN gets its UP_CHRG when reserving OUT inventory for v10?  (There is no UP_CHRG column in OUT.)

Darlene, for v12, is the ITM_INV.UP_CHRG moved to SO_LN, I assume or does GERS blank that out if OUTLET is reserved??

Thanks,
Kent

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 11, 2013, 4:07:06 PM:_

The documentation that I could find on it states that the UP_CHRG is reserved for future use, so I would guess it has absolutely no purpose. And this was in the Version 12 documentation.

Darlene

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 12, 2013, 9:45:23 AM:_

Darlene,
    We are close to sorting this out, but please help for a few more tests.

First, please look up the following three outlet pieces in both OUT and ITM_INV tables:
'1208102316', '0313225716', '1003186691'

(Recall that last time you did this, GERS automatically adjusted the mismatching costs to use the one from ITM_INV.)

```
SQL> select NAS_OUT_ID, i.itm_cd, i.store_cd, i.loc_cd, i.CST as INV_COST, o.CST as OUT_COST, i.RCV_DT, o.FIFL_DT, o.FIFO_DT
  2  from itm_inv i, out o
  3  where i.INV_STATUS = 'O' and i.NAS_OUT_ID = o.ID_CD
  4  and nas_out_id in ('1208102316','0313225716','1003186691')
  5  order by NAS_OUT_ID;

NAS_OUT_ID ITM_CD    STOR LOC_CD   INV_COST   OUT_COST RCV_DT      FIFL_DT     FIFO_DT
---------- --------- ---- ------ ---------- ---------- ----------- ----------- -----------
0313225716 890354992 11   OUTLET       51.3      51.15 01-DEC-2010 01-OCT-2010 01-NOV-2010
1003186691 382816505 11   OUTLET      37.54      57.75 01-OCT-2010 01-NOV-2010 01-DEC-2010
1208102316 467150947 12   OUTLET         67         65 01-NOV-2010 01-DEC-2010 01-OCT-2010
```

Please let me know after you've completed this first step so I can check the database.

Thanks very much!
Kent

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 14, 2013, 9:18:57 AM:_

(Changeset [[3695]](../commit/f367e254766095d362b597dbf1444fb67a960581))
* Outlet Reservation Support (closes #1958)

**Comment by kentbower** [comment]
_**kbower** trac status change on Feb 14, 2013, 9:18:57 AM:_ **closed**
